embedded/sql/engine.go

```diff
@@ -1113,100 +1215,143 @@ func (e *Engine) QueryPreparedStmt(stmt *SelectStmt, params map[string]interface
+	e.mutex.RLock()
+	defer e.mutex.RUnlock()
+
+	if e.closed {
+		return nil, ErrAlreadyClosed
+	}
 
 	// TODO (jeroiraz): won't be needed when in-memory catalog becomes transactional
 	if e.catalog == nil {
 		return nil, ErrCatalogNotReady
 	}
 
 	if renewSnapshot {
-		err := e.RenewSnapshot()
+		err := e.renewSnapshot()
 		if err != nil && err != tbtree.ErrReadersNotClosed {
 			return nil, err
 		}
 	}
 
-	snapshot, err := e.Snapshot()
+	snapshot, err := e.getSnapshot()
 	if err != nil {
 		return nil, err
 	}
 
-	implicitDB, err := e.DatabaseInUse()
+	implicitDB, err := e.databaseInUse()
 	if err != nil {
 		return nil, err
 	}
 
-	_, _, _, err = stmt.compileUsing(e, implicitDB, params)
+	_, err = stmt.compileUsing(e, implicitDB, params)
 	if err != nil {
 		return nil, err
 	}
 
 	return stmt.Resolve(e, implicitDB, snapshot, params, nil)
 }
 
-func (e *Engine) ExecStmt(sql string, params map[string]interface{}, waitForIndexing bool) (ddTxs, dmTxs []*store.TxMetadata, err error) {
+func (e *Engine) ExecStmt(sql string, params map[string]interface{}, waitForIndexing bool) (summary *ExecSummary, err error) {
 	return e.Exec(strings.NewReader(sql), params, waitForIndexing)
 }
 
-func (e *Engine) Exec(sql io.ByteReader, params map[string]interface{}, waitForIndexing bool) (ddTxs, dmTxs []*store.TxMetadata, err error) {
+func (e *Engine) Exec(sql io.ByteReader, params map[string]interface{}, waitForIndexing bool) (summary *ExecSummary, err error) {
 	stmts, err := Parse(sql)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
 	return e.ExecPreparedStmts(stmts, params, waitForIndexing)
 }
 
-func (e *Engine) ExecPreparedStmts(stmts []SQLStmt, params map[string]interface{}, waitForIndexing bool) (ddTxs, dmTxs []*store.TxMetadata, err error) {
-	e.catalogRWMux.Lock()
-	defer e.catalogRWMux.Unlock()
+type ExecSummary struct {
+	DDTxs []*store.TxMetadata
+	DMTxs []*store.TxMetadata
+
+	UpdatedRows     int
+	LastInsertedPKs map[string]uint64
+}
+
+func (e *Engine) ExecPreparedStmts(stmts []SQLStmt, params map[string]interface{}, waitForIndexing bool) (summary *ExecSummary, err error) {
+	if len(stmts) == 0 {
+		return nil, ErrIllegalArguments
+	}
+
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if e.closed {
+		return nil, ErrAlreadyClosed
+	}
 
 	if e.catalog == nil {
 		err := e.loadCatalog(nil)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 	}
 
-	implicitDB, err := e.DatabaseInUse()
-	if err != nil {
-		return nil, nil, err
+	implicitDB, err := e.databaseInUse()
+	if err != nil && err != ErrNoDatabaseSelected {
+		return nil, err
+	}
+
+	summary = &ExecSummary{
+		LastInsertedPKs: make(map[string]uint64),
 	}
 
 	for _, stmt := range stmts {
-		centries, dentries, db, err := stmt.compileUsing(e, implicitDB, params)
+		txSummary, err := stmt.compileUsing(e, implicitDB, params)
 		if err != nil {
-			return ddTxs, dmTxs, err
+			e.resetCatalog() // in-memory catalog changes needs to be reverted
```

Comment:
it will only be reloaded if some change was produced e.g. a table or index has been created, insertion with auto-incremental pk, etc.

This may change in the future, but it's not something specific to auto-increment feature itself